state/protocol/protocol_state/state/mutator.go

```diff
@@ -100,17 +92,33 @@ func newStateMutator(
 //     of the calling code (specifically `FollowerState`).
 //
 // updated protocol state entry, state ID and a flag indicating if there were any changes.
-func (m *stateMutator) Build() (hasChanges bool, updatedState *flow.ProtocolStateEntry, stateID flow.Identifier, dbUpdates []transaction.DeferredDBUpdate) {
-	updatedState, stateID, hasChanges = m.epochHappyPathStateMachine.Build()
-	dbUpdates = m.pendingDbUpdates
-	return
+func (m *stateMutator) Build() (flow.Identifier, []transaction.DeferredDBUpdate, error) {
+	var dbUpdates []transaction.DeferredDBUpdate
+	for _, stateMachine := range m.orthoKVStoreMachines {
+		dbUpdates = append(dbUpdates, stateMachine.Build()...)
+	}
+	stateID := m.kvMutator.ID()
+	version, data, err := m.kvMutator.VersionedEncode()
+	if err != nil {
+		return flow.ZeroID, nil, fmt.Errorf("could not encode protocol state: %w", err)
+	}
+
+	// Schedule deferred database operations to index the protocol state by the candidate block's ID
+	// and persist the new protocol state (if there are any changes)
+	dbUpdates = append(dbUpdates, m.kvStoreSnapshots.IndexTx(m.candidate.ID(), stateID))
```

Comment:
⚠️ I think this is wrong, because when the primary executed this logic, the `candidate` header cannot give you a valid `ID` as explained in [my comment above](https://github.com/onflow/flow-go/pull/5616#discussion_r1548696603). We are tripping over our own malleability foot-guns 😱.